---
title: "Introduction to Bioconductor class <tt>ExpressionSet</tt>"
author: "Alex Sanchez"
date: "March 29, 2020"
output:
 prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
    number_sections: true
 pdf_document:
    toc: true
    toc_depth: 2
---

```{r include=FALSE}
require(knitr)
opts_chunk$set(
concordance=FALSE, echo=TRUE, cache=TRUE, warning=FALSE, error=FALSE, message=FALSE)
```

# Introduction. Working with Omics data

Omics data are not only high throughput, what means we require big dta matrices to store raw data. They are also complex and, besides numerical data, they often require additional information such as covariates, annotations or technical information required for preprocessing the data.

In this lab we introduce the <tt>ExpressionSet</tt> class as an option for managing all these pieces of information simultaneously, which not only simplifies the process but also prevents mistakes derived from lack of consistency between the parts.

We start illustrating with a toy example what would be a \"standard approach\" to managing an omics dataset. Next we introduce the OOP paradigm and a Bioconductor class that allows encapsulating all the informations together and show how it facilitates the process of storing, managing and analyzing omics data.


## A toy dataset

For the purpose of this lab we are going to simulate a toy (fake) dataset that consists of the following:

* <b>Expression values</b> A matrix of 30 rows and 10 columns containing expression values from a gene expression experiment. Matrix column names are sample identifiers
* <b>Covariates</b> A table of ten rows and four columns containing the sample identifiers, the treatment groups and the age and sex of individuals.
\item{Genes} Information about the features contained in the data. May be the gene names, the probeset identifiers etc. Usually stored in a character vector but may also be a table with distinct annotations per feature.
* <b>Information about the experiment</b> Additional information about the study, such as the authors and their contact details or the title and url of the study that originated them.

```{r simulateData}
expressionValues <- matrix (rnorm (300), nrow=30)
colnames(expressionValues) <- paste0("sample",1:10)
head(expressionValues)
```

**VERY IMPORTANT**: To create the ExpressionSet the following has to be verified:

- The names of the columns of the object that contains the expressions, that will be stored in `assayData`
- must match the names of the rows of the object that contains the covariates, that will be stored in `phenoData`.

In this example it is saved in the variable `sampleNames` but this field will be used as the *name of the rows*, not as another column

```{r simulateCovariates}
targets <- data.frame(sampleNames = paste0("sample",1:10),
                      group=c(paste0("CTL",1:5),paste0("TR",1:5)),
                      age = rpois(10, 30), 
                      sex=as.factor(sample(c("Male", "Female"),10,replace=TRUE)),
                      row.names=1)
head(targets, n=10)
```

```{r simulateGeneInfo}
myGenes <-  paste0("gene",1:30)
```

```{r simulateInfo}
myInfo=list(myName="Alex Sanchez", myLab="Bioinformatics Lab",
          myContact="alex@somemail.com", myTitle="Practical Exercise on ExpressionSets")
show(myInfo)
```

Having data stored in this way is usually enough for most of the analyes we may want to do.
The only unconvenient comes from the fact that the information about the same individuals is in separate R objects so that, for certain applications, we will have to access several objects and <em>assume they are well related</em>.

For example if we want to make a principal components analysis and plot the groups by treatment we need to use both ``expressionValues" and ``targets."

```{r }
pcs <- prcomp(expressionValues)
names(pcs)
barplot(pcs$sdev)
plot(pcs$rotation[,1], pcs$rotation[,2], main="Representation of first two principal components")
text(pcs$rotation[,1], pcs$rotation[,2],targets$sex)
```

Or, if we sort the genes from most to least variable and whant to see which are the top variable genes. We need to use both objects ``expressionValues" and ``myGenes" assuming they are well linked:

```{r }
variab <- apply(expressionValues, 1, sd)
orderedGenes <- myGenes[order(variab, decreasing=TRUE)]
head(variab[order(variab, decreasing=TRUE)])
head(orderedGenes)
```

Imagine we are informed that individual has to be removed. We have to do it in \"expressionValues\" and \"targets\".

```{r subsetExpressions}
newExpress<- expressionValues[,-9]
newTargets <- targets[-9,]
wrongNewTargets <- targets [-10,]
```
                                                                                                            It is relatively easy to make an unnoticeable mistake in removing unrelated values from the data matrix and the targets table. If instead of removing individual 9 we remove individual 10 it may be  difficult to realize what has happened unless it causes a clear unconsistency!

Next section introduces a data structure that allows to encapsulate all these informations together ensuring that the links assumed are true.


# Bioconductor classes for omics data


## The OOP paradigm

Object-oriented design provides a convenient way to represent data structures and actions performed on them.

* A <em>class</em> can be tought of as a template, a description of what constitutes each instance of the class.
* An <em>instance</em> of a class is a realization of what describes the class.
* Attributes of a class are data components, and methods of a class are functions, or actions the instance/class is capable of.

The {\it R} language has several implementations of the OO paradigm but, in spite of its success in other languages, it is relatively minoritary.



## Bioconductor Classes

One case where OOP has succeeded in R or, at least, is more used than in others is in the Bioconductor Project ([bioconductor.org](http://bioconductor.org)). In Bioconductor we have to deal with complex data structures such as the results of a microarray experiment, a genome and its annotation or a complex multi-omics dataset. These are situations where using OOP to create classes to manage those complex types of data is clearly appropriate.


## The Biobase package

The `R`package{Biobase} package implements one of the best known Bioconductor classes: <tt>ExpressionSet</tt>. It was originally intended to contain microarray data and information on the study that generated them and it has become a standard for similar data structures.

```{r loadPackage}
require(Biobase)
```

Figure @ref(fig:ExpressionSet)~ shows the structure of this class. It is essentially a <em>container</em> that has distinct slots to store some of the most usual components in an omics dataset.

\begin{figure} <!-- ?--->
<img src=""ExpressionSet"" width="50%">
\caption{Structure of the <tt>ExpressionSet</tt> class, showing its slots and their meaning<a name="fig:ExpressionSet}"></a>
\end{figure} <!-- ?--->

The advantage of the OOP approach is that, if a new type of omics data needs a similar but different structure it can be created using inheritance, which means much less work than and better consistency than creating it from scratch.


## Creating and using objects of class ExpressionSet

In order to use a class we need to <em>instantiate</em> it, that is we need to create an object of this class.

This can be done using the generic constructor <tt>new</tt> or with the function <tt>ExpressionSet</tt>.

Both the constructor or the function require a series of parameters which roughly correspond to the slots of the class (type <tt>? ExpressionSet</tt> to see a list of compulsory and optional arguments).

In the following subsections we describe how to create an <tt>ExpressionSet</tt> using the components of our toy dataset. Some of the elements will directly be the element in the toy dataset, such as the expression  matrix. For others such as the covariates or the experiment information, specific classes have been introduced so that we have to instantiate these classes first and then use the the objects created to create the <tt>ExpressionSet</tt> object.


## Slot <tt>AssayData</tt>

The main element, and indeed the only one to be provided to create an <tt>ExpressionSet</tt>, is <tt>AssayData</tt>. For our practical purposes it can be seen as a matrix with as many rows as genes or generically \"features\" and as many columns as samples or individuals.


```{r creaExpressionSet1}
myEset <- ExpressionSet(expressionValues)
class(myEset)
show(myEset)
```



## Information about covariates

Covariates, such as those contained in the \"targets\" data frame are not included in the \"ExpressionSet\" \"as.is\". Instead we have first to create an intermediate object of class <tt>AnnotatedDataFrame</tt>.

Class `R`class{AnnotatedDataFrame} is intended to contain a data frame where we may want to provide enhanced information for columns, i.e. besides the short column names, longer labels to describe them better.

The information about covariates, contained in an instance of class <tt>AnnotatedDataFrame</tt>, is stored in the slot <tt>phenoData</tt>.

```{r AnnotatedDataFrame2}
columnDesc <-  data.frame(labelDescription= c("Treatment/Control", 
                                                "Age at disease onset", 
                                                "Sex of patient (Male/Female"))
myAnnotDF <- new("AnnotatedDataFrame", data=targets, varMetadata= columnDesc)
show(myAnnotDF)
```
Notice that we have not included a label for sample names because this information is not a column of the `phenoData` object.

Once we have an <tt>AnnotatedDataFrame</tt> we can add it to the <tt>ExpressionSet</tt>

```{r }
phenoData(myEset) <- myAnnotDF
```

Alternatively we could have created the<tt>AnnotatedDataFrame</tt> object first and then create the <tt>ExpressionSet</tt> object with both the expression values and the covariates. In this case it would be required that the expression matrix colum names are the same as the targets row names.

```{r creaEset2}
myEset <- ExpressionSet(assayData=expressionValues, phenoData=myAnnotDF)
show(myEset)
```


## Adding information about features

Similarly to what we do to store information about covariates, information about genes (or generically \"features\") may be stored in the optional slot <tt>featureData</tt> as an <tt>AnnotatedDataFrame</tt>.

The number of rows in <tt>featureData</tt> must match the number of rows in <tt>assayData.</tt> Row names of <tt>featureData</tt> must match row names of the matrix / matrices in assayData.

This slot is good if one has an annotations table that one wishes to store and manage jointly with the other values. ALternatively we can simple store the names of the features using a character vector in the slot
<tt>featureNames</tt>.

```{r }
myEset <- ExpressionSet(assayData=expressionValues,
                        phenoData=myAnnotDF,
                        featureNames =myGenes)
# show(myEset)
```

## Storing information about the experiment

In a similar way to what happens with the <tt>AnnotatedDataFrame</tt> class there has been developed a class to store information about the experiment. The structure of the class, called <tt>MIAME</tt> follows the structur of what has been described as the \"Minimum Information About a Microarray Experiment\" see [www.ncbi.nlm.nih.gov/pubmed/11726920](https://www.ncbi.nlm.nih.gov/pubmed/11726920)

This is useful information but it is clearly optional for data analysis.

```{r label=MIAME}
myDesc <- new("MIAME", name= myInfo[["myName"]],
            lab= myInfo[["myLab"]],
            contact= myInfo[["myContact"]] ,
            title=myInfo[["myTitle"]])
print(myDesc)
```

Again we could add this object to the <tt>ExpressionSet</tt> or use it when creating it from scratch.

```{r }
myEset <- ExpressionSet(assayData=expressionValues,
                        phenoData=myAnnotDF,
                        fetureNames =myGenes,
                        experimentData = myDesc)
# show(myEset)
```

## Using objects of class <tt>ExpressionSet</tt>

The advantage of working with <tt>ExpressionSets</tt> lies in the fact that action on the objects are done in such a way that its consistency is ensured. That means for instance that if we subset the <tt>ExpressionSet</tt> it is automatically done on the columns of the expressions and on the rows of the covariates and it is no possible that a distinct row/column are removed.

The following lines illustrate some management of data in an <tt>ExpressionSet</tt>.

\paragraph{Access Slot values}

 Notice that to access the values we use special fucntions called \"accessors\" instead of the dollar symbol (which would not work for classe) or the <br>@ symbol that does substitute the  \$ symbol.

 Notice also that, in order to access the data frame contained in the <tt>phenoData</tt> slot, which is an <tt>AnnotatedDataFrame</tt>, we need to use two accessors: <tt>phenoData</tt> to access the <tt>ExpressionSet</tt>'s<tt>phenoData</tt> slot and <tt>pData</tt> to access the <tt>data</tt> slot in it. It is strange until you get used to it!

```{r usingExpressionSets}
dim(exprs(myEset))
class(phenoData(myEset))
class(pData(phenoData(myEset)))
head(pData(phenoData(myEset)))
head(pData(myEset))
```

\paragraph{Subsetting <tt>ExpressionSet}</tt>

This is where the interest of using <tt>ExpressionSets</tt> is most clearly realized.

The <tt>ExpressionSet</tt> object has been cleverly-designed to make data manipulation consistent with other basic R object types. For example, creating a subset of an ExpressionsSet will subset the expression
matrix, sample information and feature annotation (if available) simultaneously in an appropriate manner. The user does not need to know how the object is represented \"under-the-hood\". In effect, we can treat the <tt>ExpressionSet</tt> as if it is a standard R data frame

```{r }
smallEset <- myEset[1:15,c(1:3,6:8)]
dim(exprs(smallEset))
dim(pData(smallEset))
head(pData(smallEset))
all(colnames(exprs(smallEset))==rownames(pData(smallEset)))
```

We can for instance create a new dataset for all individuals younger than 30 or for all females without having to worry about doing it in every component.

```{r }
youngEset <- myEset[,pData(myEset)$age<30]
dim(exprs(youngEset))
head(pData(youngEset))
```


## Exercise

1. Select a GEO dataset and prepare, from it, the components we have seen in the sections above, that is: The expression values, in a matrix or data.frame, the targets in a data frame, the experiment description, and information about annotations and gene names (you may obtain these from the matrix rownames).
2. Proceed as above and create first the pieces needed to create the <tt>ExpressionSet</tt> and then an object of class <tt>ExpressionSet</tt> with all the data and its information.
3. Reproduce the data exploration done in the first exercise accessing the data through the <tt>ExpressionSet</tt>.
4. Do some subsetting and check the consistency of the results obtained.
5. Add these steps to a new section in your "Exercise 1" document. Render the new document and when you are satisfied with it update your giyhub repository.


# The <tt>GEOquery package</tt>

## Overview of GEO

The NCBI Gene Expression Omnibus (GEO) serves as a public repository for a wide range of high-throughput experimental data. These data include single and dual channel microarray-based experiments measuring mRNA, genomic DNA, and protein abundance, as well as non-array techniques such as serial analysis of gene expression (SAGE), mass spectrometry proteomic data, and high-throughput sequencing data.

At the most basic level of organization of GEO, there are four basic entity types. The first three (Sample, Platform, and Series) are supplied by users; the fourth, the dataset, is compiled and curated by GEO staff from the user-submitted data. See the GEO home page for more information.

## Getting data from GEO

Getting data from GEO is really quite easy. There is only one command that is needed, <tt>getGEO</tt>.

This one function interprets its input to determine how to get the data from GEO and then parse the data into useful R data structures. Usage is quite simple.

To illustrate this we will use data from a study on Effect of the thermogenic gene program during adipocyte differentiation. The data for this study had been uploaded into the Gene Expression Omnibus (GEO). The dataset selected is identified with the accession number: GSE100924.

More information about the study can be found in GEO and in lab2 "Exercies in linear models and experimental design"


```{r }
if (!require(GEOquery)) {
  BiocManager::install("GEOquery")
}
require(GEOquery)
gse <- getGEO("GSE100924")
class(gse)
names(gse)
gse[[1]]
esetFromGEO <- gse[[1]]
```

The downloaded object is an <tt>ExpressionSet</tt> stored in a list. This means that instead of doing the painful process of creating the object step by step one can simply download it from GEO and start using it as in the previous section.

Notice that you can use the _extractor_ methods `exprs`, `pdata` or `annotation` to extract information such as the expression matrix, the covariates and groups or the name of the annotations.

```{r}
head(exprs(esetFromGEO))
```

Before extracting the phenodata object it is good to look at which fields it contains.

```{r}
colnames(pData(esetFromGEO))
```
Many of these columns contain repeated informations (a non-efficient way to store information about the study in the same table as the groups information).

The relevant information for the analysis seems to be in columns 38 to 41:  "cold exposure:ch1", "gender:ch1" , "genotype:ch1", "tissue:ch1".

```{r}
pData(esetFromGEO)[,c("cold exposure:ch1", "gender:ch1" , "genotype:ch1", "tissue:ch1")]
```



## Exercise


1. Use the <tt>getGEO</tt> command to create an <tt>ExpressionSet</tt> for the dataset you used in the previous exercise. Notice that the object needed is within a list so you need to access to it using the `[[]]` operator.
1. Once you have created it reproduce what you did there with your data.

